src/Plugin/InvokeStrategy/FinderInvokeStrategy.php

```diff
@@ -30,6 +30,10 @@ function (ActionEvent $actionEvent): void {
 
                 $deferred = $actionEvent->getParam(QueryBus::EVENT_PARAM_DEFERRED);
 
+                if ($actionEvent->getParam(MessageBus::EVENT_PARAM_MESSAGE_HANDLED, false)) {
```

Comment:
move this code up, finder, query and deferred don't need to get fetched before this check.